File: src/main/kotlin/com/github/dungphan/unityindex/tools/unity/GetComponentUsageTool.kt
package com.github.dungphan.unityindex.tools.unity

import com.github.dungphan.unityindex.constants.ToolNames
import com.github.dungphan.unityindex.server.models.ToolCallResult
import com.github.dungphan.unityindex.tools.AbstractMcpTool
import com.github.dungphan.unityindex.tools.schema.SchemaBuilder
import com.github.dungphan.unityindex.util.UnityAssetIndex
import com.intellij.openapi.project.Project
import kotlinx.serialization.json.JsonObject

class GetComponentUsageTool : AbstractMcpTool() {

    override val requiresPsiSync: Boolean = false

    override val name = ToolNames.GET_COMPONENT_USAGE

    override val description = """
        Find where a Unity MonoBehaviour or ScriptableObject is attached in scenes (.unity) and prefabs (.prefab). This is invisible to code-only analysis — these references are serialized in YAML asset files, not in C# code.

        Parameters:
        - typeName (required): The C# class name (e.g. "PlayerController", "EnemyAI"). Matched against .cs file names.
        - project_path (optional): Only needed with multiple projects open.

        Returns: List of asset files where the component is used, with GameObject names and file IDs.

        Example: {"typeName": "PlayerController"}
    """.trimIndent()

    override val inputSchema: JsonObject = SchemaBuilder.tool()
        .stringProperty("typeName", "The C# class name of the MonoBehaviour or ScriptableObject", required = true)
        .projectPath()
        .build()

    override suspend fun doExecute(project: Project, arguments: JsonObject): ToolCallResult {
        val typeName = requiredStringArg(arguments, "typeName")
            .getOrElse { return createErrorResult(it.message ?: "Missing typeName") }

        val index = UnityAssetIndex.create(project)
            ?: return createErrorResult("Cannot resolve Unity project directory")

        val result = index.findComponentUsages(typeName)

        if (result.scriptGuid == null) {
            return createErrorResult("No .cs script file found matching type name '$typeName'. Ensure the file is named $typeName.cs and has a .meta file.")
        }

        return createJsonResult(result)
    }
}

File: vscode-extension/src/tools/unity/getComponentUsageTool.ts
import { AbstractMcpTool, ToolContext } from "../abstractTool";
import { TOOL_NAMES } from "../../constants";
import { ToolCallResult } from "../../models/jsonRpc";
import { ProjectContext } from "../../server/projectResolver";
import { Args, requireString } from "../../utils/args";
import { SchemaBuilder } from "../../utils/schema";

export class GetComponentUsageTool extends AbstractMcpTool {
  protected readonly requiresLsp = false;
  readonly isHeavyScan = true;

  readonly name = TOOL_NAMES.GET_COMPONENT_USAGE;
  readonly description =
    "Find where a Unity MonoBehaviour or ScriptableObject is attached in scenes (.unity) and prefabs (.prefab). " +
    "These references are serialized in YAML asset files, not in C# code, so they're invisible to find_references.";
  readonly inputSchema = SchemaBuilder.tool()
    .stringProperty("typeName", "The C# class name of the MonoBehaviour or ScriptableObject", true)
    .projectPath()
    .build();

  protected async doExecute(
    project: ProjectContext,
    args: Args,
    ctx: ToolContext,
  ): Promise<ToolCallResult> {
    const typeName = requireString(args, "typeName");
    const index = await ctx.assetIndex.get(project);
    const result = await index.findComponentUsages(typeName, ctx.signal);
    if (result.scriptGuid === null) {
      return this.error(
        `No .cs script file found matching type name '${typeName}'. ` +
          "Ensure the file is named " + typeName + ".cs and has a .meta file.",
      );
    }
    return this.json(result);
  }
}
